### ISSUE TITLE
[css-compositing-2] Remove plus-darker

### ISSUE THREAD

**Issue by jakearchibald** [w3c/fxtf-drafts#447]
https://drafts.fxtf.org/compositing/#porterduffcompositingoperators_plus_darker

The definition here seems plain wrong. For instance:

```
co = max(0, 1 - αs x Cs + 1 - αb x Cb)
// Assuming
// αs = 1
// αb = 1
// Cs = 0.5
// Cb = 0
co = max(0, 1 - 1 x 0.5 + 1 - 1 x 0)
co = max(0, 1.5)
co = 1.5
```

`max(0, x)` is used, which suggests `x` will be a maximum of 1, but the definition produces values 0-2 (assuming pixel values are 0-1).

Safari supports the `plus-darker` value for `mix-blend-mode`.

The [definition in Apple's docs](https://developer.apple.com/documentation/appkit/nscompositingoperation/plusdarker) matches the spec, so it's incorrect for the same reasons. The implementation is closed source, so I can't check how it actually works.

It seems like, on WebKit Linux (via cairo), [`plus-darker` is an alias of `darken`](https://github.com/WebKit/WebKit/blob/ffef6b24a7a6aa068445259128edc473e11c2285/Source/WebCore/platform/graphics/cairo/CairoUtilities.cpp#L140), which doesn't match what I see in Safari. I might not be interpreting the source correctly.

We heard any use-cases for plus-darker, although maybe that will become clearer with correct documentation.

**Comment by pacocoursey** [comment]
@jakearchibald Would you consider advocating for `plus-darker` to be brought back into the spec and/or implementation? It's extremely useful for increased vibrancy of translucent materials, which are becoming more common on the web thanks to the power of filters and blend modes. Apple uses both `plus-lighter` and `plus-filter` blend modes across all of their operating systems, iOS in particular.

Here's an example (taken directly from [Apple's macOS Sketch Design Template](https://developers.apple.com/design/resources/)) where `plus-darker` on WebKit boosts vibrancy and legibility:

![MmkqKuKg@2x](https://user-images.githubusercontent.com/34928425/228689995-8ff5dbc4-687a-435a-80c8-2a6961340367.png)

([codepen](https://codepen.io/pxco/pen/NWLmVKr))

It also seems silly to only implement `plus-lighter` without its opposite!

**Comment by dbaron** [comment]
The first problem here is that nobody's published a formula for what Apple's `plus-darker` is that we believe is correct.

For what it's worth, if I were to try to make up a `plus-darker` formula that was the opposite of `plus-lighter` but darkening, I think I'd probably write:

Fa = 1; Fb = 1;
co = max(0, 1 - αs x Cs - αb x Cb);
αo = min(1, αs + αb);

I have not tested to see if this does anything close to reasonable.

(On a related topic, the only difference between `lighter` and `plus-lighter` appears to be the clamping to the [0, 1] range.)

**Comment by dbaron** [comment]
I wrote a test in https://github.com/web-platform-tests/wpt/pull/41713 that tests the composite modes in canvas.  It turns out my test objected to the formula in https://github.com/w3c/fxtf-drafts/issues/447#issuecomment-1489513453 because it wasn't producing properly premultiplied colors:  there were cases where the color result was greater than the alpha result, which isn't allowed.

However, my formula was sort of close; it appears (based on fiddling with my test) that what Safari implements for `plus-darker` is:

co = min(1, αs + αb) - min(1, αs x (1 - Cs) + αb x (1 - Cb));
αo = min(1, αs + αb);

(It's possible there's a simpler way to write this, but that's the expression I figured out that leads to Safari passing the test for `plus-darker`.)

**Comment by smfr** [comment]
Comments in Apple source code say:
```
//        /* R = MAX(0, 1 - ((1 - D) + (1 - S))). {R, S and D are premultiplied
//           so actually R = MAX(0, (Da+Sa) - ((Da - D) + (Sa - S))) } */
//                         = MAX(0, Da + Sa - Da + D - Sa + S
//                         = MAX(0, D + S)
```
Where R=result, D=destination, S=source as premultiplied values.

**Comment by magcius** [comment]
The logic in here is hard to follow. The original expression that CoreCG has is: `1 - ((1 - S) + (1 - D))`. This effectively works out to `S + D - 1`. It's additive, but there's an extra -1 tossed into one of the sides for good measure. That doesn't give you the generalized formula for blending with alpha, though.

> However, my formula was sort of close; it appears (based on fiddling with my test) that what Safari implements for `plus-darker` is:
> 
> co = min(1, αs + αb) - min(1, αs x (1 - Cs) + αb x (1 - Cb));
> αo = min(1, αs + αb);

I don't have an easy system to test CoreCG with, so I'm going to assume this is the actual formula. `min(1, ...)` on the left here is loadbearing, perhaps unintentionally. It's what's stopping this from becoming the same as plus-lighter. Normally, `αs + αb` would add to 2.0 for two opaque images, but since the min caps it to 1 it effectively adds the `- 1` bias on the right-hand side. Note, however, that if you have two images with alphas that don't add to more than 1, it becomes the same as plus-lighter.

There are some other formulas that you could come up with, like `S + D - Sa*Da` that I think would try to take this into account, but I can't think of any correct way to blend darker that doesn't make compromises somewhere.

> Comments in Apple source code say:
> 
> ```
> //        /* R = MAX(0, 1 - ((1 - D) + (1 - S))). {R, S and D are premultiplied
> //           so actually R = MAX(0, (Da+Sa) - ((Da - D) + (Sa - S))) } */
> //                         = MAX(0, Da + Sa - Da + D - Sa + S
> //                         = MAX(0, D + S)
> ```

This is clearly incorrect, as should be evidenced by the resulting formula being effectively the same as plus-lighter (additive blending). You can't swap out the 1 for both "Da+Sa" and "Da"/"Sa" without a clamp.

**Comment by dbaron** [comment]
That number is `47/255`.  Your formula produces that as the green component given the (non-premultiplied) inputs.  (The formulas are expected to take non-premultiplied inputs and produce premultiplied outputs.)

**Comment by heitbrink0** [comment]
> That number is `47/255`. Your formula produces that as the green component given the (non-premultiplied) inputs. (The formulas are expected to take non-premultiplied inputs and produce premultiplied outputs.)

Maybe try
co = max(0, Cb * αb - (1 - Cs) * αs)
αo = min(1, αs + αb)
